File: leetcode/94-binary-tree-inorder-traversal/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// recursive
func inorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	result := []int{}
	if root.Left != nil {
		left := inorderTraversal(root.Left)
		result = append(result, left...)
	}
	result = append(result, root.Val)
	if root.Right != nil {
		right := inorderTraversal(root.Right)
		result = append(result, right...)
	}
	return result
}

// iterative
// 1. put left child into stack recusively
// 2. when we pop a stack, set current to the pop.right such that we will not consider any node which was in the stack(avoid duplicate)
func inorderTraversal1(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	var result []int
	var stack []*TreeNode
	curr := root
	for curr != nil || len(stack) > 0 {
		// get to the left-most node
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.Left
		}
		// put the node in to the result
		pop := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		result = append(result, pop.Val)
		// consider the right node
		curr = pop.Right
	}
	return result
}

func main() {
	// 		1
	//	2		3
	// 4 5 6 7
	root := &TreeNode{1,
		&TreeNode{2,
			&TreeNode{4, nil, nil},
			&TreeNode{5, nil, nil},
		},
		&TreeNode{3,
			&TreeNode{6, nil, nil},
			&TreeNode{7, nil, nil},
		},
	}
	ans := inorderTraversal1(root)
	fmt.Println(ans)
}

File: leetcode/94-binary-tree-inorder-traversal/main.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
    recursive dfs inorder

    Time    O(n)
    Space   O(h)
"""


class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        self.res = []
        self.inorder(root)
        return self.res

    def inorder(self, node):
        if node == None:
            return
        self.inorder(node.left)
        self.res.append(node.val)
        self.inorder(node.right)


"""
    iterative dfs inorder

    Time    O(n)
    Space   O(h)
    16 ms, faster than 81.24%
"""


class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        stack = []
        cur = root
        while cur != None or len(stack) > 0:
            while cur != None:
                stack.append(cur)
                cur = cur.left
            pop = stack.pop()
            res.append(pop.val)
            cur = pop.right
        return res
